File: dataprovider-retrofit/src/main/java/ch/srg/dataProvider/integrationlayer/utils/UserAgentUtils.java
package ch.srg.dataProvider.integrationlayer.utils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;

/**
 * Copyright (c) SRG SSR. All rights reserved.
 * <p>
 * License information is available from the LICENSE file.
 */

public class UserAgentUtils {
    public static String createUserAgent(Context application) {
        String version;
        int verCode;
        try {
            PackageInfo pInfo = application.getPackageManager().getPackageInfo(application.getPackageName(), 0);
            version = pInfo.versionName;
            verCode = pInfo.versionCode;
        } catch (PackageManager.NameNotFoundException ignored) {
            version = "?";
            verCode = 0;
        }
        return application.getPackageName() + " " + version + "/" + verCode + " (" + Build.MODEL + ")";
    }
}

File: dataprovider-retrofit/src/main/java/ch/srg/dataProvider/integrationlayer/utils/UserAgentUtils.kt
package ch.srg.dataProvider.integrationlayer.utils

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build

/**
 * Copyright (c) SRG SSR. All rights reserved.
 *
 *
 * License information is available from the LICENSE file.
 */
object UserAgentUtils {
    private const val UNKNOWN_VERSION = "?"

    fun createUserAgent(application: Context): String {
        var version: String
        var verCode: Long
        try {
            val pInfo = application.packageManager.getPackageInfo(application.packageName, 0)

            @Suppress("DEPRECATION")
            val versionCode = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) pInfo.longVersionCode else pInfo.versionCode.toLong()
            version = pInfo.versionName ?: UNKNOWN_VERSION
            verCode = versionCode
        } catch (ignored: PackageManager.NameNotFoundException) {
            version = UNKNOWN_VERSION
            verCode = 0
        }
        return application.packageName + " " + version + "/" + verCode + " (" + Build.MODEL + ")"
    }
}
